server/src/mysql/mysql_writer.rs

```diff
@@ -0,0 +1,124 @@
+}
+
+impl<'a, W: std::io::Write> MysqlQueryResultWriter<'a, W> {
+    pub fn create(inner: QueryResultWriter<'a, W>) -> Self {
+        Self { inner: Some(inner) }
+    }
+
+    pub fn write(&mut self, query_result: Response) -> std::io::Result<()> {
+        if let Some(inner) = self.inner.take() {
+            return match query_result {
+                Response::AffectedRows(count) => Self::write_affected_rows(inner, count),
+                Response::Rows(rows) => Self::write_rows(inner, rows),
+            };
+        }
+        Ok(())
+    }
+
+    fn write_affected_rows(writer: QueryResultWriter<'a, W>, count: usize) -> std::io::Result<()> {
+        let res = OkResponse {
+            affected_rows: count as u64,
+            ..Default::default()
+        };
+        writer.completed(res)?;
+        Ok(())
+    }
+
+    fn write_rows(
+        writer: QueryResultWriter<'a, W>,
+        rows: Vec<HashMap<String, Datum>>,
+    ) -> std::io::Result<()> {
+        let default_response = OkResponse::default();
+        if rows.is_empty() {
+            writer.completed(default_response)?;
+            return Ok(());
+        }
+
+        fn convert_field_type(field: &Datum) -> ColumnType {
+            match field {
+                Datum::Timestamp(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::Double(_) => ColumnType::MYSQL_TYPE_DOUBLE,
+                Datum::Float(_) => ColumnType::MYSQL_TYPE_FLOAT,
+                Datum::Varbinary(_) => ColumnType::MYSQL_TYPE_VARCHAR,
+                Datum::String(_) => ColumnType::MYSQL_TYPE_VARCHAR,
+                Datum::UInt64(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::UInt32(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::UInt16(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::UInt8(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::Int64(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::Int32(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::Int16(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::Int8(_) => ColumnType::MYSQL_TYPE_LONG,
+                Datum::Boolean(_) => ColumnType::MYSQL_TYPE_SHORT,
+                Datum::Null => ColumnType::MYSQL_TYPE_VARCHAR,
+            }
+        }
+
+        fn make_column_by_filed(k: &str, v: &Datum) -> Column {
+            let column_type = convert_field_type(v);
+            Column {
+                table: "".to_string(),
+                column: k.to_string(),
+                coltype: column_type,
+                colflags: ColumnFlags::empty(),
+            }
+        }
+
+        let columns = &rows[0]
+            .iter()
+            .map(|(k, v)| make_column_by_filed(k, v))
+            .collect::<Vec<_>>();
+        let columns_keys = columns.iter().map(|c| c.column.clone()).collect::<Vec<_>>();
+        let mut row_writer = writer.start(columns)?;
+        for row in &rows {
+            for key in &columns_keys {
```

Comment:
Is it better to loop over the `columns`(no need to generate `column_keys` at L82 and clone `columns` at L83)?